src/wms/src/main/java/org/geoserver/wms/clip/CroppedGridCoverage2DReader.java

```diff
@@ -0,0 +1,104 @@
+import org.geotools.geometry.GeneralEnvelope;
+import org.geotools.geometry.jts.JTS;
+import org.geotools.referencing.CRS;
+import org.geotools.util.factory.Hints;
+import org.geotools.util.logging.Logging;
+import org.locationtech.jts.geom.Geometry;
+import org.opengis.parameter.GeneralParameterValue;
+import org.opengis.parameter.ParameterValueGroup;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+import org.opengis.referencing.operation.MathTransform;
+
+/** @author ImranR */
+public class CroppedGridCoverage2DReader extends DecoratingGridCoverage2DReader {
+    /** Parameters used to control the {@link Crop} operation. */
+    private static final ParameterValueGroup cropParams;
+
+    static final Logger LOGGER =
+            Logging.getLogger(CroppedGridCoverage2DReader.class.getCanonicalName());
+
+    static {
+        final CoverageProcessor processor =
+                new CoverageProcessor(new Hints(Hints.LENIENT_DATUM_SHIFT, Boolean.TRUE));
+        cropParams = processor.getOperation("CoverageCrop").getParameters();
+    }
+
+    /** Cached crop factory */
+    private static final Crop coverageCropFactory = new Crop();
+
+    Geometry roiGeom;
+
+    public CroppedGridCoverage2DReader(GridCoverage2DReader delegate, Geometry roiGeom) {
+        super(delegate);
+        this.roiGeom = reproject(delegate.getCoordinateReferenceSystem(), roiGeom);
+    }
+
+    private Geometry reproject(CoordinateReferenceSystem gridCRS, Geometry clipGeom) {
+        // re-project if required
+        try {
+            CoordinateReferenceSystem geomCRS = CRS.decode("EPSG:" + clipGeom.getSRID());
+            if (CRS.isTransformationRequired(geomCRS, gridCRS)) {
+                MathTransform mt = CRS.findMathTransform(geomCRS, gridCRS);
+                clipGeom = JTS.transform(clipGeom, mt);
+                clipGeom.setSRID(CRS.lookupEpsgCode(gridCRS, false));
+            }
+        } catch (Exception e) {
+            LOGGER.log(Level.SEVERE, "failed to reproject " + clipGeom.toText());
+            LOGGER.log(Level.SEVERE, e.getMessage(), e);
+        }
+
+        return clipGeom;
+    }
+
+    @Override
+    public ServiceInfo getInfo() {
+        return delegate.getInfo();
+    }
+
+    @Override
+    public ResourceInfo getInfo(String coverageName) {
+        return delegate.getInfo(coverageName);
+    }
+
+    @Override
+    public GridCoverage2D read(GeneralParameterValue[] parameters)
+            throws IllegalArgumentException, IOException {
+
+        return getCroppedGrid(super.read(parameters), roiGeom);
+    }
+
+    @Override
+    public GridCoverage2D read(String coverageName, GeneralParameterValue[] parameters)
+            throws IllegalArgumentException, IOException {
+        return getCroppedGrid(super.read(coverageName, parameters), roiGeom);
+    }
+
+    @Override
+    public GeneralEnvelope getOriginalEnvelope() {
+        return GeneralEnvelope.toGeneralEnvelope(JTS.toEnvelope(roiGeom));

```

Comment:
Shouldn't you intersect with the original envelope first? Some unit tests for this class (and the feature equivalent) would not hurt.

Reply:
understood, yes it should intersect with the result of `super.getOriginalEnvelope`